src/main/java/com/exactpro/th2/taskutils/FutureTracker.java

```diff
@@ -23,37 +23,138 @@
 
 /**
  * Following class tracks futures and when needed tries to wait for them
  */
 public class FutureTracker<T> {
+    private final ReadWriteLock lock = new ReentrantReadWriteLock();
+    private final Condition onRemove = lock.writeLock().newCondition();
+    /** Guarded by {@link #lock} */
     private final Set<CompletableFuture<T>> futures;
     private volatile boolean enabled;
+    private final int limit;
+    private final boolean hasLimit;
 
-    public FutureTracker() {
+    private FutureTracker(int limit) {
         this.futures = new HashSet<>();
         this.enabled = true;
+        this.limit = limit;
+        this.hasLimit = limit > 0;
+    }
+
+    /**
+     * Creates unlimited FutureTracker
+     * @param <T> is trackable {@link CompletableFuture} type
+     * @return instance of FutureTracker class
+     */
+    public static <T> FutureTracker<T> create() {
+        return new FutureTracker<>(0);
+    }
+
+    /**
+     * Creates FutureTracker with limit size
+     * @param <T> is trackable {@link CompletableFuture} type
+     * @param limit is max number of trackable futures if positive value, otherwise tracker unlimited
+     * @return instance of FutureTracker class
+     */
+    public static <T> FutureTracker<T> create(int limit) {
+        return new FutureTracker<>(limit);
     }
 
     /**
      * Track a future
      * @param future to be tracked
+     * @return true if future isn't done and is added for tracking otherwise false
      */
-    public void track(CompletableFuture<T> future) {
+    public boolean track(CompletableFuture<T> future) {
         if (enabled) {
             if (future.isDone()) {
-                return;
+                return false;
             }
-            synchronized (futures) {
+            lock.writeLock().lock();
+            try {
+                if (hasLimit && futures.size() == limit) {
+                    return false;
+                }
                 futures.add(future);
+            } finally {
+                lock.writeLock().unlock();
+            }
+            future.whenComplete((res, ex) -> {
+                lock.writeLock().lock();
+                try {
+                    futures.remove(future);
+                    onRemove.signalAll();
+                } finally {
+                    lock.writeLock().unlock();
+                }
+            });
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * Try to add the future for tracking during timeoutMillis
+     * @param future to be tracked
+     * @param timeoutMillis the maximum time in milliseconds to wait for free space for adding
+     * @return true if future isn't done and is added for tracking during timeout otherwise false
+     * @throws InterruptedException - if any thread has interrupted the current thread.
+     */
+    public boolean track(CompletableFuture<T> future, long timeoutMillis) throws InterruptedException {
```

Comment:
note: I think if we use `Duration` instead of long it will make the API more clear